File: python/otel_helper/processors.py
"""Custom span processors — equivalent to DebugTraceStateProcessor.cs."""

from opentelemetry.sdk.trace import SpanProcessor, ReadableSpan
from opentelemetry.trace import Span


class DebugProcessor(SpanProcessor):
    """Sets span attribute 'debug=true' on root spans when debug mode is enabled.

    The Collector tail_sampling policy 'debug-forced-attribute' uses:
        type: string_attribute
        string_attribute:
            key: debug
            values: ["true"]

    This ensures 100% sampling for debug traces regardless of language.
    """

    def on_start(self, span: Span, parent_context=None) -> None:
        if span.parent is None:
            span.set_attribute("debug", "true")

    def on_end(self, span: ReadableSpan) -> None:
        pass

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True

File: go/processors.go
package otelhelper

import (
	"context"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// debugProcessor sets debug=true attribute on root spans for Collector tail-sampling.
//
// Cross-language note: The .NET library also sets tracestate "debug=true" for Collector
// trace_state tail-sampling policy. The Go SDK ReadWriteSpan interface does NOT expose
// SetTraceState(), so we cannot set tracestate from a SpanProcessor. Python has the same
// limitation. The Collector MUST use string_attribute policy on "debug" key for
// cross-language debug trace force-sampling.
type debugProcessor struct{}

func (d *debugProcessor) OnStart(_ context.Context, s sdktrace.ReadWriteSpan) {
	if !s.Parent().IsValid() {
		s.SetAttributes(attribute.String("debug", "true"))
	}
}

func (d *debugProcessor) OnEnd(_ sdktrace.ReadOnlySpan)     {}
func (d *debugProcessor) Shutdown(_ context.Context) error   { return nil }
func (d *debugProcessor) ForceFlush(_ context.Context) error { return nil }
